ttnn/cpp/ttnn/operations/uniform/device/kernels/writer.cpp

```diff
@@ -0,0 +1,73 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "common/constants.hpp"
+#include "dataflow_api.h"
+#include "debug/dprint.h"
+
+using namespace tt;
+
+void kernel_main() {
+    uint32_t dst_addr = get_arg_val<uint32_t>(0);
+    union {
+        float f;
+        uint32_t u;
+    } f2u_from, f2u_to;
+    f2u_from.u = get_arg_val<uint32_t>(1);
+    f2u_to.u = get_arg_val<uint32_t>(2);
+    uint32_t start_id = get_arg_val<uint32_t>(3);
+    uint32_t num_tiles = get_arg_val<uint32_t>(4);
+    uint32_t end_id = start_id + num_tiles;
+
+    constexpr uint32_t cb_intermed0_id = CB::c_intermed0;
+    constexpr uint32_t cb_out0_id = CB::c_out0;
+    const InterleavedAddrGenFast<true> output_addrg = {
+        .bank_base_address = dst_addr,
+        .page_size = get_tile_size(cb_out0_id),
+        .data_format = get_dataformat(cb_out0_id)};
+
+    uint32_t max_uint = 4294967295;
+    float random_range = f2u_to.f - f2u_from.f;
+
+    for (uint32_t i = start_id; i < end_id; ++i) {
+        cb_wait_front(cb_intermed0_id, 1);
+
+        uint32_t cb_intermed0_read_ptr = get_read_ptr(cb_intermed0_id);
+        uint32_t cb_out0_write_ptr = get_write_ptr(cb_out0_id);
+
+        uint32_t *cb_intermed0_addr = reinterpret_cast<uint32_t *>(cb_intermed0_read_ptr);
+        uint8_t *cb_out0_addr = reinterpret_cast<uint8_t *>(cb_out0_write_ptr);
+
+        for (uint32_t k = 0; k < constants::TILE_WIDTH; k++) {
+            for (uint32_t j = 0; j < constants::TILE_HEIGHT; j++) {
+                uint32_t rand_uint32 = *cb_intermed0_addr;
+                float rand_float = static_cast<float>(rand_uint32) / max_uint;
+                // The hardware PRNG is not uniformly distribute.
+                // Generated rand_floats in range (0, 0.5] has higher ratio compared to (0.5, 1).
```

Comment:
```suggestion
                // Generated rand_floats in range [0, 0.5] has higher ratio compared to (0.5, 1].
```
Based on the existing explanation of the ranges `(0, 0.5]` and `(0.5, 1)`, it seems that 0 and 1 are excluded from the output range.
If 0 and 1 need to be included, modifying the ranges as suggested should address the issue.

Reply:
Looking at the implementation of PyTorch's `rand`, it includes the minimum and excludes the maximum. Please check if 1 is included in the uniform output.

Reply:
<img width="869" alt="image" src="https://github.com/user-attachments/assets/6f61b3b0-d11f-4d6e-9d29-1b7c39da43f0">
I tried on pytorch, with from=to=5. The output tensor contains only 5. So I think both from to can be included in the result.

Reply:
![image](https://github.com/user-attachments/assets/d612e624-890a-47fa-9de0-ab3405fbba26)
![image](https://github.com/user-attachments/assets/409bfe0b-4683-4ffd-ba16-424562285a3f)
According to the documentation, it seems that "exclude" is correct. The case where low equals high appears to be a special case.

Reply:
These are 2 seperate operations. I try `torch.distributions.uniform` with low = high = 5.0 and it raise an error.

<img width="928" alt="image" src="https://github.com/user-attachments/assets/638a8b65-7ef5-4e06-a981-61476d56edf5">

Which implementation should I follow `torch.Tensor.uniform_` or `torch.distributions.uniform`

Reply:
I think torch.Tensor.uniform_ is more reasonable as other operations in the epic are in torch or torch.Tensor module.

Reply:
```Python
import torch

torch.manual_seed(0)

# Tensor.uniform_ example
tensor1 = torch.empty(3, 3)
tensor1.uniform_(0, 1)

torch.manual_seed(0)
# distributions.uniform example
distribution = torch.distributions.uniform.Uniform(0, 1)
samples = distribution.sample((3, 3))

print("Tensor.uniform_ result:")
print(tensor1)
print("distributions.uniform result:")
print(samples)

----------------
Tensor.uniform_ result:
tensor([[0.4963, 0.7682, 0.0885],
        [0.1320, 0.3074, 0.6341],
        [0.4901, 0.8964, 0.4556]])
distributions.uniform result:
tensor([[0.4963, 0.7682, 0.0885],
        [0.1320, 0.3074, 0.6341],
        [0.4901, 0.8964, 0.4556]])
```
It seems that `Tensor.uniform_` and `distributions.uniform.Uniform` are the same function. 
The reason for the Python `ValueError` might be that the distributions version has one more validation step.

Reply:
![image](https://github.com/user-attachments/assets/5d300512-afd7-4eb4-8a08-fc5e9cf3f0ed)

https://github.com/pytorch/pytorch/blob/main/torch/distributions/uniform.py#L79
The sample method of distributions.uniform.Uniform calls rsample on line 79. Looking at rsample, it calls `torch.rand`, which is also defined in the range [0, 1).

Reply:
I'm still a little bit confused but since this is the behavior of torch.rand(), I will follow. Updated in this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/331cd75c78cba9765c567c53a9fcdd9921b0e9d0).